Cairo Desktop/CairoDesktop.AppGrabber/TaskCategoryProvider.cs

```diff
@@ -73,23 +79,69 @@ private void CategoryList_CollectionChanged(object sender, System.Collections.Sp
-                {
-                    appInfo = ai;
-                    break;
-                }
-                else if (window.Title.ToLower().Contains(ai.Name.ToLower()) && ai.Category != null)
+            if (TryGetCategoryDisplayNameByAppUserModelId(window, out var category)) return category;
+            if (TryGetCategoryDisplayNameByWinFileName(window, out category)) return category;
+            if (TryGetCategoryDisplayNameByProcId(window, out category)) return category;
+            if (TryGetCategoryDisplayNameByFileEquality(window, out category)) return category;
+            if (TryGetCategoryDisplayNameByTitle(window, out category)) return category;
+            return null;
+        }
+
+        private bool TryGetCategoryDisplayNameByAppUserModelId(ApplicationWindow window, out string category)
+        {
+            category = null;
+            if (!window.IsUWP) return false;
+            category = _appGrabber.CategoryList.FlatList
+                .FirstOrDefault(ai => !string.IsNullOrEmpty(ai.Target)
+                                      && window.IsUWP && ai.Target == window.AppUserModelID && ai.Category != null)
+                ?.Category.DisplayName;
+            return category != null;
+        }
+
+        private bool TryGetCategoryDisplayNameByWinFileName(ApplicationWindow window, out string category)
+        {
+            category = null;
+            if (window.IsUWP) return false;
+            category = _appGrabber.CategoryList.FlatList
+                .FirstOrDefault(ai => !string.IsNullOrEmpty(ai.Target)
+                                      && !window.IsUWP && string.Equals(ai.Target, window.WinFileName,
+                                          StringComparison.CurrentCultureIgnoreCase))
+                ?.Category.DisplayName;
+            return category != null;
+        }
+
+        private bool TryGetCategoryDisplayNameByProcId(ApplicationWindow window, out string category)
+        {
+            var applicationWindows = _shellManager.Tasks.GroupedWindows.Cast<ApplicationWindow>();
+            category = applicationWindows
+                .FirstOrDefault(ai => ai.ProcId == window.ProcId && ai.Category != null)
+                ?.Category;
+            return category != null;
+        }
+
+        private bool TryGetCategoryDisplayNameByFileEquality(ApplicationWindow window, out string category)
+        {
+            category = null;
+            if (window.IsUWP) return false;
+            var windowFileExists = File.Exists(window.WinFileName);
+            category = _appGrabber.CategoryList.FlatList
+                .FirstOrDefault(ai =>
                 {
-                    appInfo = ai;
-                }
-            }
+                    if (string.IsNullOrEmpty(ai.Target) || window.IsUWP) return false;
+                    if (!File.Exists(ai.Target) || !windowFileExists) return false;
```

Comment:
The `windowFileExists` check should happen before here since it's regarding the window rather than the ApplicationInfo. Maybe right after line 128?